tests/bluetooth/host_long_adv_recv/src/main.c

```diff
@@ -0,0 +1,482 @@
+}
+
+static void scan_timeout_cb(void)
+{
+	zassert_unreachable("Timeout should not happen");
+}
+
+static void generate_sequence(uint8_t *dest, uint16_t len, uint8_t range_start, uint8_t range_end)
+{
+	uint16_t written = 0;
+	uint8_t value = range_start;
+
+	while (written < len) {
+		*dest++ = value++;
+		written++;
+		if (value > range_end) {
+			value = range_start;
+		}
+	}
+}
+
+/* Test. */
+static void test_host_long_adv_recv(void)
+{
+	/* Register the test HCI driver */
+	bt_hci_driver_register(&drv);
+
+	/* Go! Wait until Bluetooth initialization is done  */
+	zassert_true((bt_enable(NULL) == 0), "bt_enable failed");
+
+	static struct bt_le_scan_cb scan_callbacks = { .recv = scan_recv_cb,
+						       .timeout = scan_timeout_cb };
+	bt_le_scan_cb_register(&scan_callbacks);
+	bt_addr_le_t addr_a;
+	bt_addr_le_t addr_b;
+	bt_addr_le_t addr_c;
+
+	bt_addr_le_create_static(&addr_a);
+	bt_addr_le_create_static(&addr_b);
+	bt_addr_le_create_static(&addr_c);
+
+	struct test_adv_report report_a_1 = { .length = 30, .evt_prop = MORE_TO_COME };
+	struct test_adv_report report_a_2 = { .length = 30, .evt_prop = COMPLETE };
+
+	bt_addr_le_copy(&report_a_1.addr, &addr_a);
+	bt_addr_le_copy(&report_a_2.addr, &addr_a);
+
+	struct test_adv_report report_b_1 = { .length = 30, .evt_prop = MORE_TO_COME };
+	struct test_adv_report report_b_2 = { .length = 30, .evt_prop = COMPLETE };
+
+	bt_addr_le_copy(&report_b_1.addr, &addr_b);
+	bt_addr_le_copy(&report_b_2.addr, &addr_b);
+
+	struct test_adv_report report_c = { .length = 30,
+					    .evt_prop = COMPLETE | BT_HCI_LE_ADV_EVT_TYPE_LEGACY };
+
+	bt_addr_le_copy(&report_c.addr, &addr_c);
+
+	struct test_adv_report report_a_combined = { .length = report_a_1.length +
+							       report_a_2.length };
+
+	struct test_adv_report report_a_1_repeated = { .length = CONFIG_BT_EXT_SCAN_BUF_SIZE };
+
+	struct test_adv_report report_b_combined = { .length = report_b_1.length +
+							       report_b_2.length };
+
+	generate_sequence(&report_a_combined.data[0], report_a_combined.length, 'A', 'Z');
+	generate_sequence(&report_b_combined.data[0], report_b_combined.length, 'a', 'z');
+	generate_sequence(&report_c.data[0], report_c.length, '0', '9');
+
+	(void)memcpy(&report_a_1.data[0], &report_a_combined.data[0], report_a_1.length);
+	(void)memcpy(&report_a_2.data[0], &report_a_combined.data[report_a_1.length],
+		     report_a_2.length);
+
+	for (int i = 0; i < report_a_1_repeated.length; i += report_a_1.length) {
+		memcpy(&report_a_1_repeated.data[i], &report_a_1.data[0],
+		       MIN(report_a_1.length, (report_a_1_repeated.length - i)));
+	}
+
+	(void)memcpy(&report_b_1.data[0], &report_b_combined.data[0], report_b_1.length);
+	(void)memcpy(&report_b_2.data[0], &report_b_combined.data[report_b_1.length],
+		     report_b_2.length);
+
+	/* Check that non-interleaved fragmented adv reports work */
+	ztest_returns_value(get_expected_data, &report_a_combined.data[0]);
```

Comment:
General comment: Why do `&report_a_combined.data[0]` instead of `report_a_combined.data`?

Reply:
I have seen this in other places in the codebase and I think I had some other indexes during development. Your suggestion is less noisy.